Replace RenderBuilding derivatives (RenderSimple removal Part 2)

reaperrr opened the pull request:
This replaces the four traits inheriting `RenderBuilding` with solutions based on `WithSpriteBody`.

NOTE: The `WithSpriteMakeAnimation` trait is a temporary solution that was necessary to avoid having to replace `RenderBuilding` and its four derivatives in one massive PR. It will be removed (or rather, replace the current `WithMakeAnimation`) together with `RenderBuilding`'s removal, hence why there's no upgrade rule for this.

reaperrr committed:
RenderBuildingCharge -> WithChargeAnimation

reaperrr committed:
RenderBuildingSilo -> WithSiloAnimation

reaperrr committed:
RenderBuildingTurreted -> WithTurretedSpriteBody

reaperrr committed:
RenderBuildingWall -> WithWallSpriteBody

reaperrr committed:
WithSpriteBody-based animation traits

reaperrr committed:
Extend WithSpriteBody funtionality

Move building placement range circle to PlaceBuilding, add
PauseAnimationWhenDisabled.

reaperrr commented:
Updated.
@penev92 : Adressed everything you pointed out, pushed the changes to separate commits for easier reviewing. Tell me when I can squash/fixup them.

pchote commented:
Repeating my comment from above (where it's not very obvious), I agree with @penev92's concerns about pushing too much building-specific behaviour onto `WithSpriteBody`.  Don't fall into the trap of "everything is a (must use a) WithSpriteBody"!

You could solve this in a cleaner way by having a `WithBuildingSpriteBody` trait like you suggested above, but then going a step further by moving the make animation code into this trait directly and removing the `INotifyBuildComplete` interface.  That interface has always been a horrible hack that has been prone to initialization-ordering bugs, and its use case (disabling traits until the make animation is complete) can be done more robustly using the upgrades code.

reaperrr commented:
I'm a bit at a loss, not sure how to proceed from here.

If I remove `INotifyBuildComplete` from `WithSpriteBody`, firing offsets on turrets are broken (unless the actor has WithSpriteMakeAnimation, which implements it as well).

If I move the make anim stuff to a `WithSpriteBuildingBody`, I don't know how to make facings work for traits inheriting that (in this case `WithTurretedSpriteBody`).

And I don't fully understand how exactly #7035 would work, either. How would the approach described in #7035 know about waiting for the make anim to finish if we don't implement it 'the old way'?

penev92 commented:
Short version - have a `WithMakeAnimation` that grants an upgrade when the animation is done.
Have all other traits depend on that upgrade.

reaperrr commented:
> Short version - have a WithMakeAnimation that grants an upgrade when the animation is done.
Have all other traits depend on that upgrade.

What about buildings without a make animation (d2k turrets, for example)? If that upgrade is supposed to replace INotifyBuildComplete, all buildings need it, regardless of whether they have a make animation.

pchote commented:
That isn't necessary if the upgrade is granted (to disable the traits) while the animation plays.

abcdefg30 commented:
Needs a rebase.

penev92 commented on OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs:
```diff
@@ -69,6 +72,15 @@ public string NormalizeSequence(Actor self, string sequence)
 			return RenderSprites.NormalizeSequence(DefaultAnimation, self.GetDamageState(), sequence);
 		}
 
+		public virtual void BuildingComplete(Actor self)
```
Can you change this to `AddedToWorld()`, using the `INotifyAddedToWorld` interface instead of `INotifyBuildComplete`?

penev92 commented:
OK, back to this. Sorry it has taken a while...
I think only the `INotifyAddedToWorld` is valid now that I see your last 3 commits.

reaperrr commented:
OK, will try to adress everything this weekend.

reaperrr commented:
Updated. After some cooked brain cells and discussion on IRC, I'd like to leave that "upgrades instead of INotifyBuildComplete" for later (after my render PRs are merged), as I'm struggling with the implementation anyway and don't want this PR to become any bigger.

reaperrr commented:
Rebased.

GraionDilach commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1782,6 +1782,136 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
+					}
+				}
+
+				if (engineVersion < 20150719)
```
IMO this should be redated - there was even a playtest since.

GraionDilach commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1782,6 +1782,136 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
+					}
+				}
+
+				if (engineVersion < 20150720)
```
And again.

GraionDilach commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1782,6 +1782,136 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
+					}
+				}
+
+				if (engineVersion < 20150721)
```
This too.

reaperrr commented:
Updated.

reaperrr commented:
Rebased.

Mailaender commented:
This breaks RA silos by constantly looping through all fillness states and not having auto selection sizes: 
![image](https://cloud.githubusercontent.com/assets/756669/9560840/10fbed32-4e2b-11e5-8d81-1d3c01b6d23a.png)

Mailaender commented:
This also breaks the tesla coil make animation.

Mailaender commented:
Popup turret also has no selection boundaries:

![image](https://cloud.githubusercontent.com/assets/756669/9560847/7d230f72-4e2b-11e5-885b-b201ab11cb42.png)

reaperrr commented:
Updated.
Fixed issues by reverting to `INotifyBuildComplete`, fixed selection sizes (note: these would have been fixed in #9189 at the latest, because there `AutoSelectionSize` is inherited from defaults).

Mailaender commented:
Tesla coils, obelisks and sam sites still don't have a make animation.

reaperrr commented:
Ok, it seems that splitting the removal of `RenderBuilding` in two PRs wasn't such a good idea afterall. I have trouble tracking down the make anim issues in this PR, but everything works fine in #9189, so I'm closing this in favor of the larger, but complete and more importantly now-functional #9189.
